File: coinsChange.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if coins == None or len(coins) == 0:
            return -1
        
        row = len(coins) + 1
        col = amount + 1
        dp = [[0 for i in range(col)] for j in range(row)]
        
        for j in range(col):
            dp[0][j] = amount + 1
            
        for i in range(1,row):
            for j in range(1,col):
                if j < coins[i-1]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-coins[i-1]] + 1)
        
        return -1 if dp[row-1][col-1] > amount else dp[row-1][col-1]
    
# TC: O(n*m)
# SC: O(n*m)

File: CoinsChange.java
// Time Complexity :
// Space Complexity :
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
class Solution {
    private Integer[] memo;
    public int coinChange(int[] coins, int amount) {
       int[] dp = new int[amount+1];
       Arrays.fill(dp, amount+1);

       dp[0] = 0;
       for(int i=0;i<=amount;i++){
          for(int coin:coins){
              if(i - coin < 0) continue;
              dp[i] = Math.min(dp[i],dp[i - coin] +1);
          }
       }
       return dp[amount] == (amount+1) ? -1:dp[amount];
       }
}
